keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,527 @@
+
+Taking into account the size of Kubernetes repository we will not tackle the problem of full migration to new klog methods but instead focus on improving the querability of most common logs. We will be targeting reaching a high percentage of logs in a new format generated by real cluster.
+
+### Log message structure
+
+We would like for the Kubernetes community to settle on one preferred log message structure. Proposed structure should:
+* Separate log message from its arguments
+* Treat log arguments as key-value pairs
+* Be easily parsable and queryable
+* Have a specific guidance on log message and its arguments
+
+For this purpose we suggest to use following log message structure:
+```
+<message> <key1>=<value1> <key2>=<value2> ...
+```
+
+where
+* message is formatted using `%q` fmt logic
+* keys are formatted using `%s` fmt logic
+* values are formatted using `%q` fmt logic, except for number and boolean
+* keys are separated from values with a "=” sign
+* message and key-values are separated with a single space
+
+#### New klog methods
+
+To enforce a new log message structure we want to introduce new methods to klog library which will provide a more structured interface for formatting log messages compared to current methods based on fmt format strings.
+For each format method (`Infof`, `Errorf`) we will add matching structured method (`InfoS`, `ErrorS`). Interface for this methods will model  https://github.com/go-logr/logr (suggested by @thockin)
+Each of those methods will accept log messages as a first argument and a list of key-values pairs as a variadic second argument.
+This approach allows incremental adoption of structured logging without converting ALL of kubernetes to a new API at one time.
+
+Declaration:
+
+```go
+package klog
+
+func InfoS(msg string, keysAndValues ...interface{})
+func ErrorS(err error, msg string, keysAndValues ...interface{})
+```
+
+Example:
+
+```go
+klog.InfoS("Pod status updated", "pod", "kubedns", "status", "ready")
+```
+
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kubedns" status="ready"
+```
+
+And
+
+```go
+klog.ErrorS(err, "Failed to update pod status")
+```
+
+That would result in log
+```
+E1025 00:15:15.525108       1 controller_utils.go:114] "Failed to update pod status" err="timeout"
+```
+
+### References to Kubernetes objects
+
+To address problem of inconsistency in referencing of Kubernetes objects, new klog methods will include additional logic for handling their formatting. Kubernetes objects will be recognized by checking for `ObjectMeta` structure.
+
+Namespaced objects will be represented by joining their name and namespace with `/` character. Pattern: `<namespace>/<name>`, example for pod: `kube-system/kubedns`
```

Comment:
by represented, do you mean serialized?  Do you get to choose a serialization for a log message?  What if I want the json of the object?

Reply:
Yes, we are talking about object serialization here. Currently we do not plan to support any customization of it but if there are some valid use cases when this would be useful in future we can add for instance additional configuration option for it.

Reply:
Our goal is to propose a good default. If you want to change format, you can always serialize it before passing to logging format.

Reply:
> Do we need a klog.KObj() function which is the "hint" to expand ..., "pod", klog.Kobj(pod), ... into pod/name="foo", pod/namespace="bar" ?
After suggestion from @wojtek-t we decided that using logging should not require developers to use additional methods like `Kobj` that you mentioned.
Intended objects representation should be done by default.

> Can you explain how this will actually work? Is this something that will be auto-detected? e.g. I pass in an arbitrary pointer and you test if it has metadata via reflection? Is that for any pointer? Any value? How much will that impact runtime of logging?

Our plan was to test for metadata via reflection. We will fill out the details before going into "implementable" state. We don't expect too big impact for runtime as printf operation  should dominate runtime cost.